### ISSUE TITLE
Add type safety to spec equality matcher

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#13389]
The equality matcher `eq` currently only checks the contents (via `==` semantics). It does not take into account the actual data type unless the `==` implementation validates that explicitly.

An example that came up in https://github.com/crystal-lang/crystal/pull/13108#discussion_r1118684855 is that an empty array equals any empty array, regardless of the type.

In general, the semantics of `==` are often very lax regarding type strictness. This is a wider issue discussed in #10277. Here I'm only concerned about the special case of equality matching in specs. This use case should usually have higher demands for type safety to ensure exactness of the tested interface behaviour.

Of course, with specs there are also use cases where type strictness is less important and can be safely neglected for the sake of simplicity.

So I think there is a need for both, type safe equality, and a non-type safe equality.
Of course there is also the `be_a` expectation with a similar focus on type-safety. However, it's more repetitive and adds unnecessary clutter. When the equality expectation can already take care of checking the type, there's no need to duplicate the expectation for type checking.

I did a little test run on stdlib specs and it already helped discover a bug in `Array#flatten` (#13388). I'm sure we'll find more issues like this with a more strict type checking. There were only a couple of specs that needed modification to fit the type strict expectation.

My ideal API would be `eq` (i.e. the default) to ensure type safety. As a variation, `eq!` would skip type checking.

But that is impossible to introduce because the equality semantics of `eq` are already established. Making that more strict would break lots of specs.
So my proposal would be to add `eq!` as a variant with strict type checking. The major problem with that is that this stricter and safer variant should ideally be used as default, and opting-out for less type safety should require extra effort. Now it's the other way around.

Any other ideas?

**Comment by elebow** [comment]
It's possible that a large amount of headache in the short term may be outweighed by the safety benefits in the unbounded future term. Making the "default" `eq` as safe as possible seems worthwhile, if the cost is not unbearable; users who know they can accept reduced safety can call a "non-default" function.

Separately, I'm not sure the name `eq!` is explanatory. Perhaps something like `eq`/`eq_without_type` (or `eq`/`eq_with_type`, if `eq` implies type assertion).

---

Is there a convenient way to allow gradual opt-in of type assertion in `eq`? Perhaps a compile-time flag, or a macro that can be overridden per-file? Then we could preserve existing behavior for now, make the behavior opt-out later, and eventually (or maybe never) remove it completely. And the other function `eq!` or `eq_without_type` or `eq_with_type` could always offer the other behavior.

**Comment by straight-shoota** [comment]
No, flags cannot be set from code. But a runtime class variable could work similarly to make the behaviour selectable. There's no need to know this at compile time.
I figure it could be possible to rig something up for per-file configuration, but that could easily get confusing. I'm generally a bit hesitant about allowing configuration. As long as it's for a clearly defined transitioning period, maybe it's okay.
But I think consistency is important or there will be uncertainty about what the code means.

**Comment by elebow** [comment]
I was trying to find a way like a magic comment to indicate that _this file only_ opts in to different behavior, so existing code could be migrated one file at a time. `private macro` seemed useful, but I could not find a way to do it.

**Comment by straight-shoota** [comment]
I think you could do that with a class method like `Spec.type_safe_equality!` that get's called on the top of a spec file.

That method would collect the caller locations in a list. The caller file can be optained via the pseudo constant `__FILE__` as default parameter value (https://crystal-lang.org/reference/1.8/syntax_and_semantics/constants.html#pseudo-constants). And then `#eq` would check its caller against that list.


However, I don't think this is leading to a good solution.

**Comment by HertzDevil** [comment]
Since `eq` comes from the top-level namespace via `include Spec::Expectations`, one way is to override `eq` in a file-private namespace:

```crystal
module Spec::Expectations::Strict
  macro included
    private def eq(value)
      # strict expectation
    end
  end
end
```

then include the above module in only the spec files that want to opt in:

```crystal
# spec/std/int_spec.cr
include Spec::Expectations::Strict

# spec/std/float_spec.cr
include Spec::Expectations::Strict
```

This will reveal a lot of type mismatches between number types, and commenting out either `include` disables the failures only in that file. Spec suites that want this globally can include the module inside an appropriate `spec_helper.cr`.

**Comment by straight-shoota** [comment]
For file-private implementation, instead of `macro included` we could use an explicit macro call and have it define methods directly. That way, we can prefix the macro call with a `private` modifier to define the methods only in the current scope. For global application, drop the modifier.

```cr
macro def_strict_eq
  def eq(value)
      # strict expectation
  end
end

private def_strict_eq
```

**Comment by straight-shoota** [comment]
Some more discoveries while playing with this concept:

* `eq` could be super tight to even check that the virtual types match (i.e. `typeof(value)` instead of `value.class`). This seems a bit too strict, though.
* There is a problem with negated expectations: If `subject.should_not eq value` is strict on the type, it actually loses a lot of expressive power. `1_u8.should_not eq 1_i32` might be technically true, but also quite unexpected.
* We could also consider type strictness of other expectations such as `be_close`, and `be <op>` (how about `contain`?).

**Comment by straight-shoota** [comment]
Type-strict equality expectations might not work for everything. In some cases, a loose equivalence expectation could actually be correct.

This usually concerns data types which define a natural equivalence with other representations of the same concept. For example, `1_u8` and `1_i32` describe the same number. Wrapper types like `JSON::Any` explicitly declare equivalence with their raw values. The same applies to container types containing equivalent values (`[1_u8] == [1_i32] # => true`).

The problem boils down to a distinction between _conceptual equivalence_ and _structural equality_.
I believe we need both. Crystal currently merges both under the same label `eq` and the `==` operator.
The only clear distinction is _pointer identity_ (`#same?`).

Perhaps `a.should be == b` could represent equivalence. It fits right in following the same mechanism as the comparison operators.
Only downside is that it's similar to `a.should be b` which means identity.
